### DISCUSSION TITLE
Add a config to make all fields Optional

### DISCUSSION THREAD

**Discussion by christophelec** [pydantic/pydantic#3089] [Feature Request]
# Proposal : adding a Config key to set all fields of a model as Optional

## Context :

In FastAPI, the OpenAPI spec is defined via Pydantic models.
 
To create a partial update endpoint in FastAPI, I'd like to be able to create a Model derived from another, but with all fields Optional, without having to manually duplicate all fields.

For example, let's say I have an API dealing with user, and creating a user takes as input : 

```python
class User(BaseModel):
  first_name: str
  last_name: str
  age: int
  address: str
```

If I want to have an endpoint to be able to partially update the user (e.g. change only the first_name), I'd like an endpoint where I can send :
```json
{"first_name": "NEW"}
```

To have this behavior, currently I have to either : 
* Create a specific model for patching, with all fields set as Optional : 
```python
class PatchedUser(BaseModel):
  first_name: Optional[str]
  last_name: Optional[str]
  age: Optional[int]
  address: Optional[str]
```
* Set all fields as Optional in the first model, and then add back manual validation down the road in the context of a user creation (where the fields are mandatory) (https://fastapi.tiangolo.com/tutorial/body-updates/#partial-updates-with-patch)

The first solution results in a lot of duplication, and therefore a lot of potential bugs.
The second one makes the schema less clear, and validation less easy.

## Proposed solution

I would like to be able to derive a model from this base model, but switching all the fields to Optional. A possible usage could be : 

```python
class PartialUser(User):
  class Config:
    all_optional = True
```

This way, the PartialUser would keep all the fields from User, but setting them as Optional.

The result in FastAPI would look like this : 

```python
@router.post("/users/")
async def create_user(
    user: User,
):
    # All fields are mandatory
    return create_user(user)

@router.patch("/users/{user_id}/")
async def update_user(
    user_id: int,
    user: PartialUser,
):
  # Only fields to modify will be kept
  return update_user(user.dict(exclude_unset=True))
```

I'm willing to open a PR if you think this is a good idea.

**Comment by PrettyWood** [comment]
Hello @christophelec 
As answered in the issue, you can have a look at https://github.com/samuelcolvin/pydantic/issues/2272#issuecomment-771205557.
As for the config param, I'm rather for it. The good thing is that `pydantic` supports class kwargs so it can be very close to `TypedDict`
```py
class A(TypedDict, total=False):
  x: str

class B(BaseModel, all_optional=True):
  y: str
```

**Reply by christophelec** [reply]
I did not know about the class kwargs shortcut for the config in Pydantic, pretty nice !

I did not have `total=False` in mind when thinking about the issue, but you're right that it maps well functionality wise : I'll switch to this in the PR

**Comment by anilsg** [comment]
I guess I'm a bit late to this discussion but I'm just trying to implement a partial update for PATCH against a **nested** model.

I was able to implement, (following some help on SO from others who had already solved it) a **function** that converts an **existing model** into the same model with **all fields optional**. This follows a solution from Ziur Olpa that then allows the function to be used in the path operation function as a dependency so the change to support a PATCH is simply an additional call in the path operation function without any boilerplate whatsoever: https://stackoverflow.com/a/72365032/

My implementation:

```
@lru_cache
def optional(baseclass: Type[BaseModel]) -> Type[BaseModel]:
    fields = {}
    for name, field in baseclass.__fields__.items():
        type_ = field.type_
        if type_.__base__ is BaseModel:
            fields[name] = (Optional[optional(type_)], {})
        else:
            fields[name] = (Optional[type_], None) if field.required else (type_, field.default)

    validators = {"__validators__": baseclass.__validators__}
    return create_model(baseclass.__name__ + "Optional", **fields, __validators__=validators)
```

I don't think there's normally any need for access to a named, all-optional model other than for the incoming PATCH, so I suggest this approach is cleaner than the config / class based approach.

```
@router.patch("/{obj_id}", response_model=TheBaseModel)
async def patch_document(obj_id:int, document: optional(TheBaseModel)) -> TheBaseModel:
    ....
```

I'm not sure my implementation is perfect but it seems to work ok, and deals with **nested models**. I urge you to consider nested models whatever approach you take for this request as the support for nesting doesn't seem to have been in focus on some parts of the Pydantic implementation. In particular I think the approach in the tutorial to partial updates for PATCH fails with nested models.

I haven't resolved this fully but I believe the problems I'm having when using `BaseModel.copy` for partial updates is that when the `update` parameter is supplied to the first line of the `copy` method, sub-objects are simply unpacked as `dict`s and apparently **overwrite** existing sub-objects. Therefore if a partial sub-object is supplied, the existing sub-object seems to be **replaced** rather than **updated**, effectively **erasing** any values that were missing from the partial update sub-object `dict`.

I won't dwell on this bug (?) here, as I'm still looking up some other issues which I suspect already describe this behaviour as a bug and it's not really exactly on-topic here, but, although I'm possibly too late, I just wanted to bring into focus:

1. please remember support for nested objects
2. consider the functional approach instead of, or as well as, the class approach

I concede that the function is easy to write (if my implementation is correct) and so may not be that compelling for inclusion by default.

**Comment by uriyyo** [comment]
Hi @christophelec,

You can achieve it by defining mixin class that can mark all fields as optional:
```py
from pydantic import BaseModel
from pydantic.fields import ModelField


class User(BaseModel):
    first_name: str
    last_name: str
    age: int
    address: str


class OptionalModel(BaseModel):
    class Config:
        @classmethod
        def prepare_field(cls, field: ModelField) -> None:
            field.required = False


class PartialUser(User, OptionalModel):
    pass
```

Or to make code more elegant we can define `Partial` that will transform a model into partial model:
```py
from typing import Union, TypeVar, TYPE_CHECKING, TypeAlias

from pydantic import BaseModel, BaseConfig
from pydantic.fields import ModelField


class OptionalModel(BaseModel):
    class Config(BaseConfig):
        @classmethod
        def prepare_field(cls, field: ModelField) -> None:
            field.required = False
            super().prepare_field(field)


if TYPE_CHECKING:
    T = TypeVar("T")

    Partial: TypeAlias = Union[T]
else:
    class Partial:
        def __class_getitem__(cls, item):
            return type(f"Partial{item.__name__}", (item, OptionalModel), {})


class User(BaseModel):
    first_name: str
    last_name: str
    age: int
    address: str


PartialUser = Partial[User]
```

**Reply by shanktt** [reply]
Was there some changes in Pydantic that breaks this since it was posted? When printing the model_fields all fields are still required:
```python
print(PartialUser.model_fields)
{'first_name': FieldInfo(annotation=str, required=True), 'last_name': FieldInfo(annotation=str, required=True), 'age': FieldInfo(annotation=int, required=True), 'address': FieldInfo(annotation=str, required=True)}
```

**Reply by uriyyo** [reply]
API slightly changed in v2:
```py
from typing import Any
from pydantic import BaseModel


class User(BaseModel):
    first_name: str
    last_name: str
    age: int
    address: str


class OptionalModel(BaseModel):
    @classmethod
    def __pydantic_init_subclass__(cls, **kwargs: Any) -> None:
        super().__pydantic_init_subclass__(**kwargs)

        for field in cls.model_fields.values():
            field.default = None

        cls.model_rebuild(force=True)


class PartialUser(User, OptionalModel):
    pass
```

**Reply by noamsto** [reply]
@uriyyo ,
Why do you do `cls.model_rebuild(force=True)` ?
Is it a must?
It has issue if you have lazy annotation (`ForwardRef`)
Without the `model_rebuild` it's working for me.

Here's an example this:
```python
from typing import Any
from pydantic import BaseModel


class User(BaseModel):
    first_name: str
    address: "AddressInfo"


class OptionalModel(BaseModel):
    @classmethod
    def __pydantic_init_subclass__(cls, **kwargs: Any) -> None:
        super().__pydantic_init_subclass__(**kwargs)

        for field in cls.model_fields.values():
            field.default = None

        cls.model_rebuild(force=True)


class PartialUser(User, OptionalModel):
    pass
    

class AddressInfo(BaseModel):
    city: str
    house: int
```

**Reply by uriyyo** [reply]
Yup, it's required. You can suppress this exception:
```py
class OptionalModel(BaseModel):
    @classmethod
    def __pydantic_init_subclass__(cls, **kwargs: Any) -> None:
        super().__pydantic_init_subclass__(**kwargs)

        for field in cls.model_fields.values():
            field.default = None

        with suppress(PydanticUndefinedAnnotation):
            cls.model_rebuild(force=True)
```

**Reply by noamsto** [reply]
Why is it needed?
I see it's working without `cls.model_rebuild()`, what am I missing?

BTW, a nice addition to make JsonSchema more accurate:
```python

class OptionalModel(BaseModel):
    @classmethod
    def __pydantic_init_subclass__(cls, **kwargs: Any) -> None:
        super().__pydantic_init_subclass__(**kwargs)

        for field in cls.model_fields.values():
            field.annotation = field.annotation | None  # <- for valid JsonSchema
            field.default = None

        with suppress(PydanticUndefinedAnnotation):
            cls.model_rebuild(force=True)

```

**Reply by noamsto** [reply]
@uriyyo ,
Ok I can confirm that it's working sometimes without the `model_rebuild()` and sometimes not.
So the `suppress` is needed.

**Reply by logi** [reply]
I continued building on the code from above, and my version (with newer type syntax) looks like this:
```python
class PartialModel(BaseModel):
    """Mixin to make all fields in a pydantic model optional.

    This is useful for writing RESTful PATCH endpoints where the body should take zero or more fields from the
    type being updated, and we don't want to repeat all its properties with a default.

    The resulting JSON Schema and any generated TS types will have all the expected optional bits.

    Type checkers will not understand that the properties here are optional, but in the PATCH use-case, we'll
    often just call .model_dump(exclude_unset=True) and then loop to update some attributes.

    A common way to use this is with Partial[MyModel] or NamedPartial[MyModel, "MyNewName"].
    """

    @classmethod
    def __pydantic_init_subclass__(cls, **kwargs: Any) -> None:
        super().__pydantic_init_subclass__(**kwargs)

        for field in cls.model_fields.values():
            # Make all fields optional without making them nullable
            field.default = None

        with suppress(PydanticUndefinedAnnotation):
            cls.model_rebuild(force=True)


if TYPE_CHECKING:
    # For type-checking we just pretend that Partial[T] is the same as T.
    type Partial[T] = T
    """Partial[MyModel] makes a subclass of MyModel with the OptionalModel mixin."""

    type NamedPartial[T, S] = T
    """PartFoo = NamedPartial[MyModel, "PartFoo"] makes a Partial[MyModel] with custom name in JSON schema."""
else:

    class Partial:
        def __class_getitem__(cls, item):
            name = f"Partial{item.__name__}"
            return type(name, (item, PartialModel), {})

    class NamedPartial:
        def __class_getitem__(cls, item_name):
            item, name = item_name
            return type(name, (item, PartialModel), {})
```
and you can still either mix in the `PartialModel` (renamed to keep terminology consistent) or use one of
```python
PartialThing = Partial[Thing]
RenamedPartialThing = Partial[Thing, "RenamedPartialThing"]
```

The "RenamedPartialThing" will look like a forward type reference to the type checker, and it only happens to work because that's the name of the type being defined on the same line 🙄. At runtime, however, it gets passed passed to NamedPartial which passes it to type. Don't copy that part if you don't like weird and wonderful hacks.

One significant difference is that it sets `field.default = None` to make fields optional, but does _not_ set `field.annotation = field.annotation | None` which makes the field nullable. We don't want to allow explicitly setting fields to None which should merely be optional but not nullable. This way, you can ultimately do this in your PATCH method:
```python
entity = load_entity(id)
changes = body.model_dump(exclude_unset=True)
for field, value in changes.items():
    setattr(entity, field, value)
```

FastAPI, in my case, will allow you to omit any field and the entity won't be modified. It allows you to explicitly set a nullable field to None. But it does _not_ allow you to set a non-nullable field to None. That would probably end in a DB integrity error of some sort.
